server/src/main/java/org/elasticsearch/search/aggregations/metrics/PercentilesAggregatorFactory.java

```diff
@@ -0,0 +1,126 @@
+import org.elasticsearch.search.aggregations.AggregatorFactory;
+import org.elasticsearch.search.aggregations.metrics.PercentilesAggregationBuilder.PercentilesConfig;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.AggregatorSupplier;
+import org.elasticsearch.search.aggregations.support.CoreValuesSourceType;
+import org.elasticsearch.search.aggregations.support.ValuesSource;
+import org.elasticsearch.search.aggregations.support.ValuesSourceAggregatorFactory;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.aggregations.support.ValuesSourceRegistry;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This factory is used to generate both TDigest and HDRHisto aggregators, depending
+ * on the selected method
+ */
+class PercentilesAggregatorFactory extends ValuesSourceAggregatorFactory {
+
+    private final double[] percents;
+    private final PercentilesConfig percentilesConfig;
+    private final boolean keyed;
+
+    static void registerAggregators(ValuesSourceRegistry valuesSourceRegistry) {
+        valuesSourceRegistry.register(PercentilesAggregationBuilder.NAME,
+            List.of(CoreValuesSourceType.NUMERIC, CoreValuesSourceType.HISTOGRAM),
+            new PercentilesAggregatorSupplier() {
+                @Override
+                public Aggregator build(String name, ValuesSource valuesSource, SearchContext context, Aggregator parent,
+                                        double[] percents, PercentilesConfig percentilesConfig, boolean keyed, DocValueFormat formatter,
+                                        List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) throws IOException {
+
+                    if (percentilesConfig.getMethod().equals(PercentilesMethod.TDIGEST)) {
+                        double compression = ((PercentilesConfig.TDigestConfig)percentilesConfig).getCompression();
+                        return new TDigestPercentilesAggregator(name, valuesSource, context, parent, percents, compression, keyed,
+                            formatter, pipelineAggregators, metaData);
+                    } else if (percentilesConfig.getMethod().equals(PercentilesMethod.HDR)) {
+                        int numSigFig = ((PercentilesConfig.HdrHistoConfig)percentilesConfig).getNumberOfSignificantValueDigits();
+                        return new HDRPercentilesAggregator(name, valuesSource, context, parent, percents, numSigFig, keyed,
+                            formatter, pipelineAggregators, metaData);
+                    }
+
+                    // This should already have thrown but just in case
+                    throw new IllegalArgumentException("Unknown percentiles method: [" + percentilesConfig.getMethod().toString() + "]");
```

Comment:
In the builder, we throw `IllegalStateException` for an unknown method.  Even if we never expect to hit this, we should be consistent.  I don't know which is correct, but we should pick one and use it.

Reply:
:+1: good point, hadn't noticed.  Probably stick to what the builder is throwing (at least for now) so it's consistent with how it works today.